src/Storages/MergeTree/MergeTreeIndexText.cpp

```diff
@@ -450,109 +401,199 @@ void MergeTreeIndexGranuleText::analyzeDictionary(MergeTreeIndexReaderStream & s
-            return condition_text.dictionaryBlockCache()->getOrSet(
-                TextIndexDictionaryBlockCache::hash(state.part.getDataPartStorage().getFullPath(), state.index.getFileName(), state.index_mark, block_id),
-                load_dictionary_block);
-
-        return load_dictionary_block();
+        auto hash = TextIndexDictionaryBlockCache::hash(state.part.getDataPartStorage().getFullPath(), state.index.getFileName(), block_id);
+        return condition_text.dictionaryBlockCache()->getOrSet(hash, load_dictionary_block);
     };
 
     for (const auto & [block_idx, tokens] : block_to_tokens)
     {
         const auto dictionary_block = get_dictionary_block(block_idx);
+
         for (const auto & token : tokens)
         {
-            auto it = remaining_tokens.find(token);
-            chassert(it != remaining_tokens.end());
             auto * token_info = dictionary_block->getTokenInfo(token);
 
             if (token_info)
             {
-                ProfileEvents::increment(ProfileEvents::TextIndexBloomFilterTruePositives);
-                it->second = *token_info;
+                remaining_tokens.emplace(token, *token_info);
             }
-            else
+            else if (global_search_mode == TextSearchMode::All)
             {
-                ProfileEvents::increment(ProfileEvents::TextIndexBloomFilterFalsePositives);
+                remaining_tokens.clear();
+                return;
+            }
+        }
+    }
+}
 
-                if (global_search_mode == TextSearchMode::All)
-                {
-                    remaining_tokens.clear();
-                    return;
-                }
+PostingListPtr MergeTreeIndexGranuleText::readPostingsBlock(
+    MergeTreeIndexReaderStream & stream,
+    MergeTreeIndexDeserializationState & state,
+    const TokenPostingsInfo & token_info,
+    size_t block_idx)
+{
+    auto * data_buffer = stream.getDataBuffer();
 
-                remaining_tokens.erase(it);
-            }
+    const String & data_path = state.part.getDataPartStorage().getFullPath();
+    const String & index_name = state.index.getFileName();
+    const auto & condition_text = assert_cast<const MergeTreeIndexConditionText &>(*state.condition);
+
+    const auto load_postings = [&]() -> PostingListPtr
+    {
+        ProfileEvents::increment(ProfileEvents::TextIndexReadPostings);
+        stream.seekToMark({token_info.offsets[block_idx], 0});
+        return PostingsSerialization::deserialize(*data_buffer, token_info.header, token_info.cardinality);
+    };
+
+    auto hash = TextIndexPostingsCache::hash(data_path, index_name, token_info.offsets[block_idx]);
+    return condition_text.postingsCache()->getOrSet(hash, load_postings);
+}
+
+void MergeTreeIndexGranuleText::readPostingsForRareTokens(MergeTreeIndexReaderStream & stream, MergeTreeIndexDeserializationState & state)
+{
+    for (const auto & [token, token_info] : remaining_tokens)
+    {
+        if (token_info.embedded_postings)
```

Comment:
In l. 467 and l. 472, it seems slightly more idiomatic to check if `EmbeddedPostings` and `SingleBlock` are set.